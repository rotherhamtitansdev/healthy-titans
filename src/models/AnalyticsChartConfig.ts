import { NHSNutritionalDataModel } from "./NHSNutritionalDataModel";

// This is the numerical instance of the NHS Nutritional guidelines model
export const NHSNutritionalBoundaries: NHSNutritionalDataModel[] = [
  { name: "Fat", low: 3, high: 17.5 },
  { name: "Saturates", low: 1.5, high: 5 },
  { name: "Sugars", low: 5, high: 22.5 },
  { name: "Salt", low: 0.1, high: 0.6 },
];

// These are the config parameters for the chart styling and sizing
export const chartSizing = {
  chartHeight: 500,
  barWidth: 30,
  CStartPos: 50,
  RStartPos: 60,
  circleYDif: 100,
  smallerRounded: 15,
};

export enum ChartLabel {
  Calories = "ENERC_KCAL",
  Protein = "PROCNT",
  Fat = "FAT",
  Saturates = "FASAT",
  Sugars = "SUGAR",
  Salt = "NA",
}

export enum Colour {
  Green = 1,
  Yellow = 2,
  Red = 3,
}
