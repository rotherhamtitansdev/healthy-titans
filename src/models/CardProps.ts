export interface CardProps {
  name: string;
  additionalStyling?: string;
  onClick?: () => void;
}
