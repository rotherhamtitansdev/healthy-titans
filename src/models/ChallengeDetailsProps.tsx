export interface ChallengeDetailsProps {
  name: string;
  description: string;
  img: string;
  rules: string[];
  equipment: string[];
  howToPlay?: string[];
}
