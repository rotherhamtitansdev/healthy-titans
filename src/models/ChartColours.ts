export interface ChartColours {
  name: string;
  primary: string;
  secondary: string;
  status: number;
  position: number;
}
