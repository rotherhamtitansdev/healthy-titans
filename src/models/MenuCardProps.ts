export interface MenuCardProps {
  key: number;
  name: string;
  description?: string;
  img?: string;
  path: string;
  externalPath?: boolean;
  disableOnClick?: boolean;
  firebaseName?: string;
  additionalStyling?: string;
}
