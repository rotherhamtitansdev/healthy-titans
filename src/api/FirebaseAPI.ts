import { getDownloadURL, getStorage, ref } from "firebase/storage";
import {
  collection,
  doc,
  getDoc,
  getDocs,
  query,
  where,
  documentId,
} from "firebase/firestore";
import { FoodDetailsProps } from "../data/nutritional_information/FoodDetailsComponentData";
import { fStore } from "../config/firebase-config";
import { NutritionalDetailsFirebaseProps } from "../models/NutritionDetailsComponentData";

interface SeeNextProps {
  key: number;
  name: string;
  path: string;
  img: string;
}

class FirebaseAPI {
  static fetchImages = async (firebaseName: string): Promise<string> => {
    const storage = getStorage();
    const starsRef = ref(storage, firebaseName);
    return getDownloadURL(starsRef);
  };

  static fetchAllImages = async (firebaseNames: string[]) =>
    Promise.all(firebaseNames.map((firebaseName) => FirebaseAPI.fetchImages(firebaseName)));

  static fetchFoodDetailsComponentsData = async (): Promise<
    FoodDetailsProps[] | undefined
  > => {
    const querySnapshot = await getDocs(collection(fStore, "FYPData"));
    if (!querySnapshot) return undefined;
     return querySnapshot.docs.map((each) => each.data() as FoodDetailsProps);
  };

  static fetchFoodDetailsSingle = async (name: string) => {
    const docRef = doc(fStore, "FYPData", name);
    const docSnap = await getDoc(docRef);
    if (docSnap) return docSnap.data() as FoodDetailsProps;
    return undefined;
  };

  static fetchFoodDetailsSeeNext = async (
    category: string,
    currentName: string
  ): Promise<{ cardData: FoodDetailsProps[]; paths: string[] }> => {
    const q = query(
      collection(fStore, "FYPData"),
      where("category", "==", category),
      where("name", "!=", currentName)
    );
    const docSnap = await getDocs(q);
    const docs = docSnap.docs.map((data) => {
      const newDoc = data.data();
      newDoc.path = data.id;
      return newDoc;
    });

    const shuffled = docs.sort(() => 0.5 - Math.random()).slice(0, 3);

    return {
      cardData: shuffled.map((value) => ({
        name: value.name,
        description: value.description,
        firebaseName: value.firebaseName,
        category: value.category,
        score: value.score,
        facts: value.facts,
      })),
      paths: shuffled.map((value) => value.path),
    };
  };

  static fetchNutritionData = async (name: string): Promise<NutritionalDetailsFirebaseProps> => {
    const querySnapshot = await getDocs(collection(fStore, "NutritionData", name, "Content"));

    const sorted = querySnapshot.docs.sort((a, b) => {
      if (a.data().order > b.data().order) {
        return 1;
      }
      if (b.data().order > a.data().order) {
        return -1;
      }
      return 0;
    });

    return sorted.map((doc2) => {
      const data = doc2.data();
      return Object.entries(data).map(([key,value]) => ({key,value}));
    });
  };

  static fetchNutritionSeeNext = async (currentName: string): Promise<SeeNextProps[]> => {
    const q = query(collection(fStore, "NutritionData"), where(documentId(), "!=", currentName));
    const querySnapshot = await getDocs(q);

    const docs = querySnapshot.docs.sort(() => 0.5 - Math.random()).slice(0, 3);

    return Promise.all(
      docs.map(async (sortedDoc, index) => {
        const dataDoc = await getDoc(doc(fStore, "NutritionData", sortedDoc.id, "Content", "Main"));

        const URI = await FirebaseAPI.fetchImages(dataDoc.data()?.firebaseName);

        return { key: index, name: dataDoc.data()?.name, path: docs[index].id, img: URI };
      })
    );
  };

  static fetchQuizData = async () => {
    const docRef = doc(fStore, "QuizData", "Quiz");
    const quizDoc = await getDoc(docRef);

    if (quizDoc.exists()) {
      return quizDoc.data();
    }
    return undefined;
  };

  static fetchAboutUsData = async () => {
    const docRef = doc(fStore, "AboutUsData", "Data");
    const aboutUsDoc = await getDoc(docRef);

    if (aboutUsDoc.exists()) {
      return aboutUsDoc.data();
    }
    return undefined;
  };

  static fetchChallengesData = async (currentName: string) => {
    const docRef = doc(fStore, "Challenges", currentName);
    const challengesDoc = await getDoc(docRef);

    if (challengesDoc.exists()) {
      return challengesDoc.data();
    }
    return undefined;
  };

  static fetchChallengesList = async () => {
    const querySnapshot = await getDocs(collection(fStore, "ChallengesData"));

    if (!querySnapshot) return undefined;
    return querySnapshot.docs.map((each) => each.data());
  };
}

export default FirebaseAPI;
