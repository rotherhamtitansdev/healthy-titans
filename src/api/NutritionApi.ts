import axios from "axios";
import { NutritionDetailsAPIModel } from "../models/NutritionDetailsAPIModel";

const getNutritionalDetailsSingular = async (food:string):Promise<NutritionDetailsAPIModel> => {
  const url = "https://api.edamam.com/api/nutrition-details?app_id=558ec9f4&app_key=d33fe22a6604c71d5459842dd5030eb3";
  const data = {
    title: food,
    ingr: [
      `100g ${food}`,
    ],
  };

  // Return is of type NutritionDetailsAPIModel
  const res = await axios.post<NutritionDetailsAPIModel>(url, data, {
    headers: {
      "Content-Type": "application/json",
    },
  });
  return res.data;
};

export default getNutritionalDetailsSingular;
