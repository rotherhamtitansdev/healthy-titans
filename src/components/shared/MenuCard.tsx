import React, { useCallback, useEffect, useState } from "react";
import "../../App.css";
import { useNavigate } from "react-router-dom";
import { ClipLoader } from "react-spinners";
import { MenuCardProps } from "../../models/MenuCardProps";
import Card from "./Card";
import Image from "./Image";
import FirebaseAPI from "../../api/FirebaseAPI";
import openInNewTab from "../../functions/Navigate";

/*
 * This component represents a card used in a menu screen
 * */
const MenuCard = (props: MenuCardProps) => {
  const [getImage, setImage] = useState<React.ReactNode | undefined>();

  const navigate = useNavigate();
  const navigateToPath = useCallback(() => navigate(props.path, { replace: false }), [navigate]);

  useEffect(() => {
    if (props.firebaseName !== undefined) {
      FirebaseAPI.fetchImages(props.firebaseName).then((URI) =>
        setImage(<Image img={URI} alt={props.name} />)
      );
    } else if (props.img !== undefined) {
      setImage(<Image img={props.img} alt={props.name} />);
    }
  }, []);

  const handleOnClick = () => {
    if (!props.disableOnClick) {
      return props.externalPath ? () => openInNewTab(props.path) : () => navigateToPath();
    }
    return () => undefined;
  };

  return (
    <Card
      card={{
        name: props.name,
        additionalStyling: `${props.additionalStyling}`,
        onClick: handleOnClick(),
      }}
    >
      <div className="py-[1.75rem] md:py-[3.625rem] w-full h-full flex-col flex">
        <div className="w-full h-full flex justify-center items-center md:-mt-0 md:-mt-8">
          {getImage || <ClipLoader size={100} color="#DC476D" className="" />}
        </div>
        <div className="md:text-xl text-homepageHeaderText font-semibold text-center mt-auto">
          {props.name}
        </div>
      </div>
    </Card>
  );
};

export default MenuCard;
