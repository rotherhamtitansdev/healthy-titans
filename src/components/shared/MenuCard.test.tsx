import React from "react";
import { render, screen, waitFor } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { MemoryRouter } from "react-router";
import * as FirebaseAPI from "../../api/FirebaseAPI";
import MenuCard from "./MenuCard";
import { MenuCardProps } from "../../models/MenuCardProps";

const MockMenuCard = {
  key: 0,
  name: "Mock Card 1",
  path: "/mock-path-1",
  firebaseName: "MockImages/mock-card-1.jpg",
} as MenuCardProps;

beforeEach(() => {
  jest.spyOn(FirebaseAPI, "fetchImages").mockResolvedValue("fruit.jpg");
});

describe("Menu Card", () => {
  test("component renders with image", async () => {
    const { asFragment } = render(
      <MenuCard
        key={MockMenuCard.key}
        name={MockMenuCard.name}
        path={MockMenuCard.path}
        externalPath={MockMenuCard.externalPath}
        firebaseName="MockImages/mock-card-1.jpg"
        additionalStyling=""
      />,
      {
        wrapper: MemoryRouter,
      }
    );

    expect(await screen.findByRole("img", { name: "" })).toHaveAttribute("src", "fruit.jpg");
    expect(asFragment()).toMatchSnapshot();
  });

  test("component renders with no firebase", async () => {
    render(
      <MenuCard
        key={MockMenuCard.key}
        name={MockMenuCard.name}
        img="fruit.jpg"
        path={MockMenuCard.path}
        externalPath={MockMenuCard.externalPath}
        additionalStyling=""
      />,
      {
        wrapper: MemoryRouter,
      }
    );

    expect(await screen.findByRole("img", { name: "" })).toHaveAttribute("src", "fruit.jpg");
  });

  test("component renders and can click card", async () => {
    const user = userEvent.setup();

    render(
      <MenuCard
        key={MockMenuCard.key}
        name={MockMenuCard.name}
        img={MockMenuCard.img}
        path={MockMenuCard.path}
        externalPath={MockMenuCard.externalPath}
        firebaseName={MockMenuCard.firebaseName}
        additionalStyling=""
      />,
      {
        wrapper: MemoryRouter,
      }
    );

    user.click(await screen.findByRole("button", { name: "Mock Card 1" }));
    waitFor(() => expect(window.location.pathname).toBe("/mock-path-1"));
  });

  test("component renders and can not click card", async () => {
    const user = userEvent.setup();

    render(
      <MenuCard
        key={MockMenuCard.key}
        name={MockMenuCard.name}
        img={MockMenuCard.img}
        path={MockMenuCard.path}
        externalPath={MockMenuCard.externalPath}
        firebaseName={MockMenuCard.firebaseName}
        disableOnClick
        additionalStyling=""
      />,
      {
        wrapper: MemoryRouter,
      }
    );

    user.click(await screen.findByRole("button", { name: "Mock Card 1" }));
    waitFor(() => expect(window.location.pathname).toBe("/"));
  });
});
