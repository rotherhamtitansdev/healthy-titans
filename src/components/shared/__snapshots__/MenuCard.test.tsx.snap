// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Menu Card component renders with image 1`] = `
<DocumentFragment>
  <div
    class="bg-white font-quicksand table align-middle rounded-3xl shadow-lg  hover:cursor-grab hover:-translate-y-0.5 hover:scale-105 duration-300"
    data-testid="Mock Card 1"
    role="button"
    tabindex="0"
  >
    <div
      class="py-[1.75rem] md:py-[3.625rem] w-full h-full flex-col flex"
    >
      <div
        class="w-full h-full flex justify-center items-center md:-mt-0 md:-mt-8"
      >
        <div
          class="px-[2.75rem] md:px-[5rem] pb-[0.5rem] md:pb-[1rem] min-w-[89px] min-h-[81px] 2xl:mt-5"
        >
          <img
            alt=""
            class="mx-auto undefined"
            draggable="false"
            src="fruit.jpg"
          />
        </div>
      </div>
      <div
        class="md:text-xl text-homepageHeaderText font-semibold text-center mt-auto"
      >
        Mock Card 1
      </div>
    </div>
  </div>
</DocumentFragment>
`;
